Creating react game

jiajdo opened the pull request.

jiajdo committed:
Added .gitignore and set up react assets

jiajdo committed:
Added keyboard and board components. Modified CSS file.

jiajdo committed:
Imported react in board and keyboard components

jiajdo committed:
Added Words helper file

jiajdo committed:
Moved Words.jsx to components folder. Fixed import syntax typos

jiajdo committed:
Moved useState(boardDefault) into App. Fixed boardDefault import syntax

jiajdo committed:
Comment on words.jsx

jiajdo committed:
Added createContext hooks and modified index.css

jiajdo committed:
Rearranged elements

dbritto-dev commented on gameweek6/src/components/Letter.jsx:
```diff
@@ -0,0 +1,15 @@
+    const { board } = useContext(AppContext)
+    const letter = board[attemptVal][letterPosition]
+  return (
+    <div className='letter'> </div>
```
```suggestion
    <div className='letter'>{letter}</div>
```

jiajdo committed:
Keyboard set up. Added letter expression in Letter func

jiajdo committed:
Added key.jsx component

jiajdo committed:
Centered board and keyboard components

jiajdo committed:
Added useState for attempts and selectLetter onClick function

jiajdo committed:
Set up condition for if player presses ENTER

jiajdo committed:
Added onSelect, onEnter, onDelete functions

Functions are in App and called with conditions in Key

jiajdo committed:
Added functionality for using physical keyboard to type

Added useCallback and useEffect hooks. Set conditions for keydown event to match what is being typed.

jiajdo committed:
Added conditions for showing if word attempt is correct

Set conditions to change cell color for when a letter is correct, almost correct, or wrong
Also added wordle-bank.txt

jiajdo committed:
Created generateWordSet async func and useEffect hook for the promise

jiajdo committed:
Added disabledLetter useState and added to useEffect hook

jiajdo committed:
Modified useEffect hook for disabledLetter state

Added disabledLetter id expression to key divs for page's keyboard

jiajdo committed:
Added GameOver component

jiajdo committed:
Added gameOver state and conditions

jiajdo committed:
Added toUpperCase in correctWord conditional check

jiajdo committed:
Fixed gameOver conditionals

eatnumber1 commented on gameweek6/src/App.jsx:
```diff
@@ -0,0 +1,120 @@
+    for (let i=0; i<5; i++){
+      currentWord += board[currentAttempt.attempt][i].toLowerCase()
+      console.log(currentWord)
+    console.log(correctWord)
```
Fix indentation (otherwise it looks like this statement is not part of the loop)

jiajdo replied:
Done

eatnumber1 commented on gameweek6/src/components/Key.jsx:
```diff
@@ -0,0 +1,27 @@
+function Key({ keyVal, bigKey, disabled }) {
+    const { onSelectLetter, onDelete, onEnter  } = useContext(AppContext)
+
+   //Grabs functions from App and sets conditions for pressing Enter, Delete, or keyboard letters
```
Fix indentation (add a space)

jiajdo replied:
Done

eatnumber1 commented on gameweek6/src/components/Key.jsx:
```diff
@@ -0,0 +1,27 @@
+    }
+    return (
+        //   use keyVal prop in div and set conditional id for bigKeys to apply CSS conditionally
+        <div className="key" id={bigKey ? "big": disabled && 'disabled'} onClick={selectLetter}>
```
The associativity rules of the ternary operator (`? :`) and the logical and (`&&`) often have surprising behavior. I think it mostly works, but consider the set of possible cases:

1. `true ? "big": false && 'disabled'` -> `'big'`
1. `true ? "big": true && 'disabled'` -> `'big'`
1. `false ? "big": true && 'disabled'` -> `'disabled'`
1. `false ? "big": false && 'disabled'` -> `false`

In particular, it's case 4 here that I don't think you're intending.

Usually you'd do two things:

1. Use parentheses to express your intent, e.g. `bigKey ? "big": (disabled && 'disabled')`, and
2. Replace your `disabled && 'disabled'` with another ternary expression that explicitly says what to do if `disabled` is false. E.g. `bigKey ? "big": (disabled ? 'disabled' : 'enabled')`

jiajdo replied:
Done

eatnumber1 commented on gameweek6/src/components/Keyboard.jsx:
```diff
@@ -0,0 +1,78 @@
+    }, [handleKeyboard])
+
+    return (
+        <div className='keyboard' onKeyDown={handleKeyboard}>
```
The way you've written this works, but as an alternative, consider:

Instead of duplicating this code for each row of the keyboard, instead try putting the keyboard in a two-dimensional array. E.g.

```
    const keys = [
        ['Q', 'W', 'E', 'R', 'T', 'Y', 'U', 'I', 'O', 'P'],
        ['A', 'S', 'D', 'F', 'G', 'H', 'J', 'K', 'L'],
        ['Z', 'X', 'C', 'V', 'B', 'N', 'M']
    ]
```

Then here, nest two calls to map to render the keyboard.

jiajdo replied:
Seems like the alternative would work, but this is already written, is short, and it works.

eatnumber1 commented on gameweek6/src/components/Letter.jsx:
```diff
@@ -0,0 +1,33 @@
+    //checks if current attempt is greater than maximum attempts allowed and sets conditions for
+    //CSS id of letter cell so cell can turn the appropriate color
+    const letterState = currentAttempt.attempt > attemptVal &&
+    (correct ? 'correct' : almost ? 'almost' : 'error') 
```
Here as in the other file, it's a bit hard to tell what's going on when you nest ternary operators without parentheses. I'd suggest `(correct ? 'correct' : (almost ? 'almost' : 'error'))`.

Also, what's your intent for letterState to be if `currentAttemtn.attempt <= attemptVal`? That case isn't covered.

However here (unlike in your html in the other file), using a ternary operator at all is hurting readability. I'd recommend writing this as a series of if statements instead. E.g.

```
var letterState = 'error'
if (currentAttempt.attempt > attemptVal) {
  if (correct) {
    letterState = 'correct'
  } else if (almost) {
    letterState = 'almost'
  }
}
```
